File: app/javascript/channels/messages_channel.js
// app/javascript/channels/messages_channel.js
import consumer from "./consumer"

document.addEventListener('turbolinks:load', () => {
  const messagesElement = document.getElementById('messages');
  if (messagesElement) {
    const chatId = messagesElement.dataset.chatId;

    const subscription = consumer.subscriptions.create(
      { channel: "MessagesChannel", chat_id: chatId },
      {
        connected() {
          console.log("Connected to MessagesChannel for chat " + chatId);
        },

        disconnected() {
          console.log("Disconnected from MessagesChannel");
        },

        received(data) {
          const parser = new DOMParser();
          const doc = parser.parseFromString(data, 'text/html');
          const message = doc.body.firstChild;
          messagesElement.appendChild(message);
        },

        speak(message) {
          this.perform('speak', { message: message });
        }
      }
    );

    const form = document.querySelector('form');
    form.addEventListener('submit', event => {
      event.preventDefault();
      const messageInput = form.querySelector('textarea');
      subscription.speak(messageInput.value);
      messageInput.value = '';
    });
  }
});

File: app/channels/messages_channel.rb
# app/channels/messages_channel.rb
class MessagesChannel < ApplicationCable::Channel
  def subscribed
    stream_from "chat_#{params[:chat_id]}"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def speak(data)
    message = Message.create!(
      content: data['message'],
      user_id: current_user.id,
      chat_id: params[:chat_id]
    )
    ActionCable.server.broadcast("chat_#{params[:chat_id]}", render_message(message))
  end

  private

  def render_message(message)
    ApplicationController.render(
      partial: 'messages/message',
      locals: { message: message }
    )
  end
end
